### PR TITLE
Implements Tethys Reactpy App Scaffold

### PR DESCRIPTION
**Pull request by shawncrawley**
### Description
This merge request implements a Tethys ReactPy App scaffold. I will flesh out this description soon with all of the ins and outs of it. I just wanted to finally get this in as a draft pull request to get more eyes on it.

### Changes Made to Code:
 - 

### Related
- 

### Additional Notes
-  

### Quality Checks
 - [] New code is 100% tested
 - [] Code has been formated
 - [] Code has been linted
 - [] Docstrings for new methods have been added

### CODE REVIEW

File: environment.yml
Comment (sdc50) [inline environment.yml:24]: Is this valid syntax for a conda package? I though that was a `pip` thing.

File: tethys_apps/base/controller.py
Comment (sdc50) [inline tethys_apps/base/controller.py:40]: Why is this an `optional_import`?  Since it's part of the source code we should be able to import it directly right?

File: environment.yml
Comment (shawncrawley) [reply environment.yml:24]: Hm... It must be, since I installed tethys fresh from this branch on a new machine a couple of days ago and it worked without issue.

File: tethys_apps/base/controller.py
Comment (shawncrawley) [reply tethys_apps/base/controller.py:40]: I debated a bit on how I should implement this. It's optional because the tethys_apps.base.page_handler module requires that `reactpy` (the true optional import) be imported at the top level since one of the importable functions itself has a `reactpy` decorator. So importing anything from `tethys_apps.base.page_handler` would fail if `reactpy` was not installed. But I just dug into it again, and realized that I actually updated the `reactpy` import in `tethys_apps.base.page_handler` itself to have optional behavior. So I think you're right that we ultimately don't need the optional behavior here anymore. I'll get that cleaned up.

File: tethys_apps/base/url_map.py
Comment (sdc50) [inline tethys_apps/base/url_map.py:73]: Is this only being used by the ReactPy stuff, or is it/can it be leveraged more broadly?

File: tethys_apps/templatetags/site_settings.py
Comment (sdc50) [inline tethys_apps/templatetags/site_settings.py:17]: Does this work on Windows? If a file path is absolute on Windows it will start with something like `C:\` right?

Comment (sdc50) [inline tethys_apps/templatetags/site_settings.py:17]: I'm not sure that this is the right move here. I'm trying to decide if the user should specify these as a URL path or a system path.

File: tethys_apps/templatetags/site_settings.py
Comment (sdc50) [inline tethys_apps/templatetags/site_settings.py:27]: In this case these are definitely URL paths so this approach won't work. To ensure we are supporting windows we'll need to ensure these are output as Posix Paths.

File: tethys_cli/install_commands.py
Comment (sdc50) [inline tethys_cli/install_commands.py:859]: I would prefer this to be called something like `script` rather than `command`. Since it is a path to a file I think `command` is somewhat missleading.

File: tethys_cli/scaffold_templates/app_templates/reactpy/install.yml_tmpl
Comment (sdc50) [inline tethys_cli/scaffold_templates/app_templates/reactpy/install.yml_tmpl:16]: I'd love to see this get put on `conda-forge`

File: tethys_cli/scaffold_templates/app_templates/reactpy/setup.py_tmpl
Comment (sdc50) [inline tethys_cli/scaffold_templates/app_templates/reactpy/setup.py_tmpl:1]: I think we're going to need to move to `pyproject.toml` sooner that later. Should we just start this scaffold out that way so we don't have to migrate it later?

File: tethys_portal/settings.py
Comment (sdc50) [inline tethys_portal/settings.py:223]: channels is not needed?

**Review by sdc50 — COMMENTED** [review]
@shawncrawley This is amazing and I'm really excited to get it officially in Tethys. I have several questions (see inline). 

In general I would prefer that `pathlib.Path` be used rather than `os.path`, but I won't hold this up for that if you don't have the bandwidth to change that now.

File: tethys_apps/base/url_map.py
Comment (shawncrawley) [reply tethys_apps/base/url_map.py:73]: This is only being used for ReactPy stuff. I did not put any though into whether it could be used more broadly. This was implemented specifically for use in the automatic generation of the navigation bar links - one per function decorated by `@page`.

File: tethys_cli/install_commands.py
Comment (shawncrawley) [reply tethys_cli/install_commands.py:859]: Yeah, I'm honestly not sure why I even switch to `command` from `path_to_post`. I think I'll just revert to `path_to_post`.

File: tethys_cli/scaffold_templates/app_templates/reactpy/post_install.py
Comment (shawncrawley) [reply tethys_cli/scaffold_templates/app_templates/reactpy/post_install.py:1]: Oh wow, that's a great idea. I would much prefer that, I just didn't even think of it.

File: tethys_cli/scaffold_templates/app_templates/reactpy/setup.py_tmpl
Comment (shawncrawley) [reply tethys_cli/scaffold_templates/app_templates/reactpy/setup.py_tmpl:1]: That makes sense to me. I don't know much about pyproject.toml, but my guess is it'd be a quick learn, yeah? I'll make it happen.

File: tethys_portal/settings.py
Comment (shawncrawley) [reply tethys_portal/settings.py:223]: According to the latest documentation, it appears it is not necessary unless using Channel's runworker command (see [here](https://channels.readthedocs.io/en/stable/installation.html)). However, this likely isn't the case on the previous versions of Django. Since including it seems to be optional now, we should probably include it anyway to get backwards compatibility.

File: tethys_apps/templatetags/site_settings.py
Comment (sdc50) [reply tethys_apps/templatetags/site_settings.py:27]: This is what we should be using: `urllib.parse.ulrjoin` https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin

**Comment by shawncrawley** [comment]
@sdc50 I implemented your feedback. In full disclosure, I got a little carried away when swapping out os.path with pathlib.Path. I ended up removing os.path from the entire project, except for about two spots where pathlib.Path was actually more clunky or problematic. So this PR is definitely super gold-plated by scope creep at this point. That means that there are a lot of new modified files since your last review... I apologize in advance. I was already in too deep before I determined that I shouldn't have added all of that to this specific PR.

File: tests/unit_tests/test_tethys_apps/test_base/test_workspace.py
Comment (sdc50) [inline tests/unit_tests/test_tethys_apps/test_base/test_workspace.py:60]: Doesn't `TethysWorkspace` accept a `Path` object?

File: tests/unit_tests/test_tethys_apps/test_template_loaders.py
Comment (sdc50) [inline tests/unit_tests/test_tethys_apps/test_template_loaders.py:69]: no need to use `Path` twice here.

File: tethys_cli/cli_helpers.py
Comment (sdc50) [inline tethys_cli/cli_helpers.py:44]: Why not use `Path` here?

File: tests/unit_tests/test_tethys_apps/test_base/test_workspace.py
Comment (sdc50) [reply tests/unit_tests/test_tethys_apps/test_base/test_workspace.py:60]: So `TethysWorkspace` will accept a `Path`. With your changes it just converts it to a `Path` first think anyway. However, it's not worth changing since this works fine and the workspaces API is going to be dropped in 5.0.

File: tests/unit_tests/test_tethys_apps/test_template_loaders.py
Comment (sdc50) [inline tests/unit_tests/test_tethys_apps/test_template_loaders.py:69]:
```suggestion
                str(Path.home() / "foo" / "template1" / "foo"),
```

File: tethys_cli/cli_helpers.py
Comment (sdc50) [inline tethys_cli/cli_helpers.py:44]:
```suggestion
    manage_path = Path(get_tethys_src_dir()) / "tethys_portal" / "manage.py"
```

**Review by sdc50 — CHANGES_REQUESTED** [review]
@shawncrawley just a couple of minor changes (I've added suggestions that you can just commit if you agree with them).

I'm very excited about the reactpy addition. The `Path` changes are all internal, but everything is converted back to a string in return values. I think this is the right call for now (to ensure we aren't breaking anything). However, in Tethys 5.0 I think we should switch and return `Path` objects wherever possible.

File: tethys_cli/cli_helpers.py
Comment (shawncrawley) [reply tethys_cli/cli_helpers.py]: Good catch! I'll fix that.

**Comment by shawncrawley** [comment]
@swainn @sdc50 I've got everything done and passing tests except oddly the "Tethys-CI / Conda Build (ubuntu-latest)" test run. Any thoughts there on what could be going on? I have reactpy-django now installed by default, which did throw a wrench in some of the testing. Maybe it's related here? But oddly only this one build...

**Comment by shawncrawley** [comment]
@gagelarsen Do you have any idea on the failing "Conda Build" GitHub workflow?

File: tethys_cli/scaffold_templates/app_templates/reactpy/pyproject.toml_tmpl
Comment (swainn) [inline tethys_cli/scaffold_templates/app_templates/reactpy/pyproject.toml_tmpl:1]: 💖 Update the other scaffolds to use this pyrpoject.toml too, please.

Comment (swainn) [reply tethys_cli/scaffold_templates/app_templates/reactpy/pyproject.toml_tmpl:1]: I think you need to add something like this to include non-python files in the package build:

```
[tool.setuptools]
include-package-data = true

[tool.setuptools.packages.find]
include = ["tethysapp*"]

[tool.setuptools.package-data]
"*" = [
    "*.js",
    "*.png",
    "*.gif",
    "*.jpg",
    "*.html",
    "*.css",
    "*.gltf",
    "*.json",
    "*.svg",
]
```

**Review by swainn — CHANGES_REQUESTED** [review]

File: environment.yml
Comment (shawncrawley) [reply environment.yml:1]: Should I still list "reactpy" separately and distinctly, even though it's baked in as an underlying requirement of "reactpy-django"?

File: environment.yml
Comment (shawncrawley) [reply environment.yml:110]: I've been looking into that. Unfortunately, it looks like there may be 4 total packages we'd need to add, since all of the dependencies also need to be in conda-forge. Those 4 packages are: reactpy-django, reactpy, reactpy-router, and nodejs-bin...

Comment (swainn) [reply environment.yml:110]: That's pretty typical for adding conda packages. I think you might be able to add them all in one go if they are related...

File: tethys_sdk/components/utils.py
Comment (shawncrawley) [reply tethys_sdk/components/utils.py:1]: Yeah, we could just leave it as event. In the cases that I used it, I would end up wanting `event` to be an argument of some of the functions used surrounding it. I just didn't want any namespace confusion. But it would make more sense to rename my function args rather than the decorator.

File: tests/unit_tests/test_tethys_apps/test_base/test_app_base.py
Comment (shawncrawley) [reply tests/unit_tests/test_tethys_apps/test_base/test_app_base.py:1561]: I just wanted a mock-like instance of UrlMap... I should have just used a mock.MagicMock object, lol. I'll swap it for that.

File: environment.yml
Comment (shawncrawley) [reply environment.yml:111]: I went ahead and added `reactpy` to the dependencies

File: environment.yml
Comment (shawncrawley) [reply environment.yml:1]: I went ahead and added `reactpy` to the dependencies.

File: tests/unit_tests/test_tethys_apps/test_base/test_app_base.py
Comment (shawncrawley) [reply tests/unit_tests/test_tethys_apps/test_base/test_app_base.py:1561]: Turns out I must've chosen `Namespace` for a reason. Using `mock.MagicMock` causes the tests to fail because mocking the property `name` cannot be done on initialization (see [mock docs](https://docs.python.org/3/library/unittest.mock.html#mock-names-and-the-name-attribute)). While I could have broken it out into two steps, it seemed like using `Namespace` shouldn't be a big deal. Let me know if you disagree.

File: tethys_cli/scaffold_templates/app_templates/reactpy/pyproject.toml_tmpl
Comment (shawncrawley) [reply tethys_cli/scaffold_templates/app_templates/reactpy/pyproject.toml_tmpl:1]: Done

**Comment by shawncrawley** [comment]
It looks like the only thing that remains is getting `reactpy` and `reactpy-django` added to conda-forge. I found that we'd also have to add `reactpy-router`, which unfortunately doesn't have an sdist file (required, I believe, by conda-forge). But I reached out to the reactpy maintainers and they quickly responded that they'll get the sdist file added by tomrorow. Once that's done, I should be able to submit the PR to the [conda staged-recipes](https://github.com/conda-forge/staged-recipes) repo and get those added. It's the final stretch!

File: tests/unit_tests/test_tethys_apps/test_base/test_app_base.py
Comment (swainn) [reply tests/unit_tests/test_tethys_apps/test_base/test_app_base.py:1561]: Try using the `UrlMaps` for the test instead of the `Namespace`.

**Review by swainn — APPROVED** [review]
Barring conda package and all comments addressed.

**Pull request merged**
